Generate native binary

datho7561 opened the pull request:
Use GraalVM's `native-image` feature to generate a executable. To test, install GraalVM and the native-image tool, then set JAVA_HOME to the GraalVM installation. Run `./mvnw package -Dnative -DskipTests` in order to generate an executable under `org.eclipse.lemminx/target`.

Includes a GitHub Action in order to run builds of the binary on PRs for testing purposes.

Adopted from [PR 673](#673)
A part of #314.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 commented:
Recently rebased against master and fixed XML formatting (it had a reflection error)

fbricon commented:
Opening the attached xsd with the binary server, I ran into a stackoverflow
[AUTOSAR_00047.xsd.zip](https://github.com/eclipse/lemminx/files/5158050/AUTOSAR_00047.xsd.zip)
```
java.lang.StackOverflowError
	at com.oracle.svm.core.graal.snippets.StackOverflowCheckSnippets.newStackOverflowError0(StackOverflowCheckImpl.java:370)
	at com.oracle.svm.core.graal.snippets.StackOverflowCheckSnippets.newStackOverflowError(StackOverflowCheckImpl.java:366)
	at com.oracle.svm.core.graal.snippets.StackOverflowCheckSnippets.throwNewStackOverflowError(StackOverflowCheckImpl.java:346)
	at com.oracle.svm.core.graal.jdk.ArraycopySnippets.doArraycopy(ArraycopySnippets.java:83)
	at java.util.ArrayList.add(ArrayList.java:484)
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
        ... 10045 more lines
	at java.util.ArrayList$SubList.add(ArrayList.java:1057)
	at java.util.AbstractList.add(AbstractList.java:108)
	at org.eclipse.lemminx.services.XMLFoldings.getFoldingRanges(XMLFoldings.java:81)
	at org.eclipse.lemminx.services.XMLLanguageService.getFoldingRanges(XMLLanguageService.java:222)
	at org.eclipse.lemminx.XMLTextDocumentService.lambda$foldingRange$10(XMLTextDocumentService.java:319)
	at java.util.concurrent.CompletableFuture.biApply(CompletableFuture.java:1119)
	at java.util.concurrent.CompletableFuture$BiApply.tryFire(CompletableFuture.java:1084)
	at java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:457)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:163)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
```

to be fair, the JVM server becomes unresponsive when opening the file with the default memory settings

datho7561 commented:
> Opening the attached xsd with the binary server, I ran into a stackoverflow

I just tried opening the file, and I couldn't reproduce the stackoverflow. The server remained responsive. Folding was automatically disabled, though.

datho7561 commented:
Moved the native binary pipeline into a separate file for the time being, as it was blocking Jenkins builds on other PRs

datho7561 commented:
Switched to building under a container in Linux

datho7561 commented:
Update on the progress of the binary:
 * The next step for the CI/CD is to notarize the app
   * Under macOS, this means connecting a Apple Developer account
 * We still need to figure out where the released binaries are going to be stored

rgrunber commented:
As LemMinX is a project at Eclipse, have you considered signing on Eclipse infrastructure ? Although they generally sign jars, I see https://www.eclipse.org/cbi/sitedocs/eclipse-macsigner-plugin/sign-mojo.html seems to be supported.

datho7561 commented:
> As LemMinX is a project at Eclipse, have you considered signing on Eclipse infrastructure ? 

This is probably the best option in the long run. The currently blocker for building on the Eclipse infrastructure is that we don't have access to macOS and Windows Jenkins agents in order to build the binaries for these platforms. I think the goal is to eventually move the builds to the Eclipse infrastructure, though.

rgrunber commented:
https://wiki.eclipse.org/Jenkins#Can_I_add_a_dedicated_agent_to_my_Jenkins_instance.3F implies it should just be a matter of opening a bug and maybe getting a reference to agent once permitted. I also see some Jenkins instances that restrict where the project can be run to the label of the agent they need.

datho7561 commented:
Rebased; symbols broken ATM...

datho7561 commented:
Fixed symbols, now rebasing...

datho7561 commented:
Rebased... needs more thorough testing to find any reflection exceptions

datho7561 commented on .github/workflows/native-image.yaml:
```diff
@@ -0,0 +1,40 @@
+    runs-on: windows-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: ilammy/msvc-dev-cmd@v1.3.0
```
Need to update this Action because there is a vulnerability

fbricon commented on org.eclipse.lemminx/src/main/resources/META-INF/native-image/resource-config.json:
```diff
@@ -0,0 +1,34 @@
+    {"pattern":"\\Qschemas/xsd/XMLSchema.xsd\\E"},
+    {"pattern":"\\Qschemas/xsd/datatypes.dtd\\E"},
+	{"pattern":"\\Qschemas/xsd/xml.xsd\\E"},
+	{"pattern":"\\Qorg/eclipse/lemminx/services/snippets/catalog-snippets.json\\E"},
```
`pattern":"org/eclipse/lemminx/services/snippets/*$"` might work, see https://www.graalvm.org/reference-manual/native-image/Resources/

Might work for resources embedded in dependency jars too

datho7561 replied:
I modified it to use a regex for the snippet JSON files. What do you mean by using it for resources in dependency jars?

datho7561 commented on README.md:
```diff
@@ -61,6 +61,34 @@ this.forward(clientConnection, serverConnection)
+		- Settings are all deserialized, so whenever a setting is added, this must be done and tested
+	- Manually test the binary and check the logs for reflection errors/NPEs
+- Snippets:
+	- If you are creating a new snippet file, make sure to register it under `org.eclipse.lemminx/src/main/resources/META-INF/resource-config.json`
```
I'll remove this as it is handled by as regex

datho7561 commented on .editorconfig:
```diff
@@ -0,0 +1,16 @@
+root = true
```
I will remove this file

rgrunber reviewed (commented):
Looks good to me. On Fedora I just needed to install just glibc-static, libstdc++-static, and zlib-static .

I noticed that there is a separate profile to use the static libraries on Linux. The only difference I'm seeing is in the additional --static flags. Is it possible to have one profile that gets activated on linux and sets the normally empty property to --static, and just refer to that property in the profile that makes the plugin call ?

If not possible, this is fine as well.

datho7561 commented:
I'll take a look into doing this.

datho7561 commented:
I changed it so that you compile with `./mvnw clean package -Dnative -DskipTests -Dgraalvm.static=--static` under Linux in order to include the libraries statically, and updated the GitHub Action in order to reflect this. Please let me know if this makes sense.

rgrunber commented:
> I changed it so that you compile with `./mvnw clean package -Dnative -DskipTests -Dgraalvm.static=--static` under Linux in order to include the libraries statically, and updated the GitHub Action in order to reflect this. Please let me know if this makes sense.

Yup, this is fine.

datho7561 commented:
Should I go ahead and merge this PR?

fbricon commented:
go ahead
